src/main/java/com/sps/Servlets/NewRecipeServlet.java

```diff
@@ -0,0 +1,146 @@
+    Collection<EmbeddedEntity> steps = getParameters(request, STEP, null);
+
+    Entity recipe = new Entity("Recipe");
+    recipe.setProperty("name", name);
+    recipe.setProperty("description", description);
+    recipe.setProperty("tags", tags);
+    recipe.setProperty("ingredients", ingredients);
+    recipe.setProperty("steps", steps);
+    recipe.setProperty("search-strings", new ArrayList<String>(searchStrings));
+    datastore.put(recipe);
+
+    response.sendRedirect("/edit-recipe.html");
+  }
+
+  /**
+   * Gets the parameters for fields that have different numbers of parameters from recipe to recipe.
+   * For example, one recipe may have 2 ingredients, while another may have 20.
+   * This method ensures that all tags, ingredients, and steps are recorded, no matter how many of each a recipe has.
+   */
+  private Collection<EmbeddedEntity> getParameters(HttpServletRequest request, String type, Collection<String> searchStrings) {
+    Collection<EmbeddedEntity> parameters = new LinkedList<>();
+    int parameterNum = 1;
+
+    String parameterName = type + parameterNum;
+    String parameter = request.getParameter(parameterName);
+    while (parameter != null) {
+      EmbeddedEntity parameterEntity = new EmbeddedEntity();
+      parameterEntity.setProperty(type, parameter);
+      addToSearchStrings(searchStrings, parameter);
+      parameterName = type + (++parameterNum);
+      parameter = request.getParameter(parameterName);
+      parameters.add(parameterEntity);
+    }
+    return parameters;
+  }
+
+  /**
+   * Adds a formatted search string to the set of search strings.
+   */
+  private void addToSearchStrings(Collection<String> searchStrings, String stringToAdd) {
+    if (searchStrings == null) {
+      return;
+    }
+    searchStrings.add(stringToAdd.toUpperCase());
+  }
+
+  private Recipe entityToRecipe(Entity recipeEntity) {
+    String name = (String) recipeEntity.getProperty("name");
+    String description = (String) recipeEntity.getProperty("description");
+    LinkedList<String> tags = (LinkedList<String>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("tags"), TAG);
+    LinkedList<String> ingredients = (LinkedList<String>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("ingredients"), INGREDIENT);
+    LinkedList<Step> steps = (LinkedList<Step>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("steps"), STEP);
+    return new Recipe(name, description, tags, ingredients, steps);
+  }
+
+  private LinkedList<Object> getDataAsList(Object propertiesObject, String type) {
+    Collection<EmbeddedEntity> properties = (Collection<EmbeddedEntity>) propertiesObject;
+    LinkedList<Object> dataAsList = new LinkedList<>();
+    for (EmbeddedEntity property : properties) {
+      dataAsList.add(property.getProperty(type));
+    }
+    return dataAsList;
+  }
+
+  private String convertToJsonUsingGson(Recipe recipe) {
+    Gson gson = new Gson();
```

Comment:
You could alternatively create a static field gson and use it instead of creating it every time.